a few functional tests

tomnis opened the pull request:
- check that a semantic rule rewrites code on disk
- check that a semantic rule in "checkScalafix" will throw an exception and leave code unmodified
- check that a syntactic rule violation will throw an exception
- check that excluded files that violate rules will not throw an exception
- a bit of README clarification that i found confusing when using the plugin in my projects

marcelocenerine commented on src/test/groovy/io/github/cosmicsilence/scalafix/ScalafixPluginFunctionalTest.groovy:
```diff
@@ -13,22 +14,22 @@ class ScalafixPluginFunctionalTest extends Specification {
-    mavenCentral()
-}
-'''
+        buildFile.write ScalafixTestConstants.BUILD_SCRIPT
```
let's keep the build script inlined here (instead of a constant) and append the other bits that you correctly introduced (Scala dependency + compiler option) 👍

marcelocenerine commented on src/test/groovy/io/github/cosmicsilence/scalafix/ScalafixPluginFunctionalTest.groovy:
```diff
@@ -13,22 +14,22 @@ class ScalafixPluginFunctionalTest extends Specification {
+        scalaSrcFile = new File(scalaSrcDir.absolutePath +  "/HelloWorld.scala")
+        scalaSrcFile.write ScalafixTestConstants.FLAWED_SCALA_CODE
+
+        scalafixConfFile = testProjectDir.newFile(ScalafixExtension.DEFAULT_CONFIG_FILE)
```
it would be nice to have a utility method that took the content of the source file and created it at the appropriate location so we could play with different issues, multiple files and also with files that have no issue. Something along these lines:

```groovy
    private File createSourceFile(String content) {
        def scalaSrcDir = testProjectDir.newFolder("src", "main", "scala")
        def sourceFile = new File(scalaSrcDir.absolutePath,  "package.scala")
        sourceFile.write content
        sourceFile
    }
```

and then we can use in individual tests like so:

```groovy
    def 'scalafix should run semantic rule that fixes scala code'() {
        given:
        def inputSourceFile = createSourceFile('''
import scala.util.Random
object HelloWorld
''')
        scalafixConfFile.write "rules = [ RemoveUnused ]"

        when:
        runGradleTask('scalafix', [ ])

        then:
        inputSourceFile.getText() == '''
object HelloWorld
'''
    }
```

tomnis replied:
cool i will inline the build script and add this `createSourceFile` method

marcelocenerine commented on src/test/groovy/io/github/cosmicsilence/scalafix/ScalafixPluginFunctionalTest.groovy:
```diff
@@ -151,6 +152,80 @@ sourceSets {
         buildResult.output.contains(':checkScalafixIntegTest SKIPPED')
     }
 
+    def 'scalafix semantic rule RemoveUnused fixes scala code'() {
```
sug.: 'scalafix should run semantic rewrite rule and fix input source files'

marcelocenerine commented on src/test/groovy/io/github/cosmicsilence/scalafix/ScalafixPluginFunctionalTest.groovy:
```diff
@@ -151,6 +152,80 @@ sourceSets {
+        scalaSrcFile.getText() == ScalafixTestConstants.FIXED_SCALA_CODE
+    }
+
+    def 'checkScalafix semantic rule RemoveUnused fails a build'() {
```
sug.: 'checkScalafix should run semantic rewrite rule and fails the build without fixing input source files'

marcelocenerine commented on src/test/groovy/io/github/cosmicsilence/scalafix/ScalafixPluginFunctionalTest.groovy:
```diff
@@ -151,6 +152,80 @@ sourceSets {
+        // code is not changed on disk
+        scalaSrcFile.getText() == ScalafixTestConstants.FLAWED_SCALA_CODE
+        final UnexpectedBuildFailure err = thrown()
+        err.message.contains("A file on disk does not match the file contents if it was fixed with Scalafix")
```
👍

marcelocenerine commented on src/test/groovy/io/github/cosmicsilence/scalafix/ScalafixPluginFunctionalTest.groovy:
```diff
@@ -151,6 +152,80 @@ sourceSets {
+    }
+
+
+    def 'scalafix semantic rule RemoveUnused skips excluded scala files'() {
```
sug.: 'scalafix should skip excluded source files'

Maybe we should have the same test for `checkScalafix`

marcelocenerine commented on src/test/groovy/io/github/cosmicsilence/scalafix/ScalafixPluginFunctionalTest.groovy:
```diff
@@ -151,6 +152,80 @@ sourceSets {
+        scalaSrcFile.getText() == ScalafixTestConstants.FLAWED_SCALA_CODE
+    }
+
+    def 'scalafix syntactic rule DisableSyntax.noVars fails a build'() {
```
- sug.: 'scalafix should run syntactic linter rule and fails the build if any violation is reported'
- it would be nice to set `autoConfigureSemanticdb` to false in tests where we run only syntactic rules

marcelocenerine commented on src/test/groovy/io/github/cosmicsilence/scalafix/ScalafixPluginFunctionalTest.groovy:
```diff
@@ -151,6 +152,80 @@ sourceSets {
+        then:
+        scalaSrcFile.getText() == ScalafixTestConstants.FLAWED_SCALA_CODE
+        final UnexpectedBuildFailure err = thrown()
+        err.message.contains("error: [DisableSyntax.var] mutable state should be avoided")
```
we could add this too:

`err.message.contains "A linter error was reported"`

marcelocenerine commented on src/test/groovy/io/github/cosmicsilence/scalafix/ScalafixPluginFunctionalTest.groovy:
```diff
@@ -151,6 +152,80 @@ sourceSets {
+        then:
+        scalaSrcFile.getText() == ScalafixTestConstants.FLAWED_SCALA_CODE
+        final UnexpectedBuildFailure err = thrown()
+        err.message.contains("error: [DisableSyntax.var] mutable state should be avoided")
```
same comments from the previous test apply here

marcelocenerine commented on src/test/groovy/io/github/cosmicsilence/scalafix/ScalafixTestConstants.groovy:
```diff
@@ -0,0 +1,49 @@
+/**
+ * Created by tomas.mccandless on 2/12/20.
+ */
+class ScalafixTestConstants {
```
as suggested in other comments, I think this class could go away if we inlined the build script and created source files as we need them in the individual tests

marcelocenerine commented on src/test/groovy/io/github/cosmicsilence/scalafix/ScalafixPluginFunctionalTest.groovy:
```diff
@@ -151,6 +152,80 @@ sourceSets {
+        err.message.contains("error: [DisableSyntax.var] mutable state should be avoided")
+    }
+
+
```
it would be nice to have a test for the happy path: a source file without any issues :)

tomnis commented on src/main/groovy/io/github/cosmicsilence/scalafix/ScalafixExtension.groovy:
```diff
@@ -7,7 +7,7 @@ import org.gradle.api.provider.SetProperty
 class ScalafixExtension {
 
-    private static final String DEFAULT_CONFIG_FILE = ".scalafix.conf"
+    protected static final String DEFAULT_CONFIG_FILE = ".scalafix.conf"
```
i made this change before #9 was put in. if you like i can change this back to private and modify the tests to pass rules via command line instead of writing out a conf file

marcelocenerine replied:
it is ok to write the file. Some rules require additional configuration (e.g. `DisableSyntax`) which can only be provided via config file

marcelocenerine replied:
we have the `.scalafix.conf` string repeated in some tests already. It is not a big deal if you need to add another one.

tomnis replied:
ok, cool. didnt realize that was repeated already. i will change it back to private

tomnis commented on src/test/groovy/io/github/cosmicsilence/scalafix/ScalafixPluginFunctionalTest.groovy:
```diff
@@ -185,6 +415,20 @@ sourceSets {
+     * @return Written file containing the provided source code content.
+     */
+    private File createSourceFile(String content) {
+        final File scalaSrcDir = testProjectDir.newFolder("src", "main", "scala", "io", "github", "cosmicsilence", "scalafix")
```
i use a full package path here so we can easily test file exclusion

tomnis commented:
hi @marcelocenerine i have addressed your feedback, please take another look when you get a chance. cheers.

@rooeque @sarcasticsimba just looping yall in, in case you are curious
